bundles/org.openhab.binding.cbus/src/main/java/org/openhab/binding/cbus/internal/CBusHandlerFactory.java

```diff
@@ -0,0 +1,99 @@
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerFactory;
+import org.openhab.binding.cbus.CBusBindingConstants;
+import org.openhab.binding.cbus.handler.CBusCGateHandler;
+import org.openhab.binding.cbus.handler.CBusDaliHandler;
+import org.openhab.binding.cbus.handler.CBusLightHandler;
+import org.openhab.binding.cbus.handler.CBusNetworkHandler;
+import org.openhab.binding.cbus.handler.CBusTemperatureHandler;
+import org.openhab.binding.cbus.handler.CBusTriggerHandler;
+import org.openhab.binding.cbus.internal.discovery.CBusGroupDiscovery;
+import org.openhab.binding.cbus.internal.discovery.CBusNetworkDiscovery;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * The {@link CBusHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Scott Linton - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = ThingHandlerFactory.class, configurationPid = "binding.cbus")
+public class CBusHandlerFactory extends BaseThingHandlerFactory {
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return CBusBindingConstants.SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Nullable
+    @Override
+    protected ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (thingTypeUID.equals(CBusBindingConstants.BRIDGE_TYPE_CGATE)) {
+            CBusCGateHandler handler = new CBusCGateHandler((Bridge) thing);
+            registerDeviceDiscoveryService(handler);
+            return handler;
+        }
+
+        if (thingTypeUID.equals(CBusBindingConstants.BRIDGE_TYPE_NETWORK)) {
+            CBusNetworkHandler handler = new CBusNetworkHandler((Bridge) thing);
+            registerDeviceDiscoveryService(handler);
+            return handler;
+        }
+
+        if (thingTypeUID.equals(CBusBindingConstants.THING_TYPE_LIGHT)) {
+            return new CBusLightHandler(thing);
+        }
+
+        if (thingTypeUID.equals(CBusBindingConstants.THING_TYPE_TEMPERATURE)) {
+            return new CBusTemperatureHandler(thing);
+        }
+
+        if (thingTypeUID.equals(CBusBindingConstants.THING_TYPE_TRIGGER)) {
+            return new CBusTriggerHandler(thing);
+        }
+
+        if (thingTypeUID.equals(CBusBindingConstants.THING_TYPE_DALI)) {
+            return new CBusDaliHandler(thing);
+        }
+
+        return null;
+    }
+
+    private void registerDeviceDiscoveryService(CBusCGateHandler cbusCgateHandler) {
```

Comment:
You also need to handle the removal of ThingHandlers.
You can look at org.openhab.binding.bluetooth.internal.BluetoothHandlerFactory for reference.

Reply:
@cpmeister Are you talking about implementing removeHandler? 
I had a quick look at BluetoothHandlerFactory but dont see anything there that is significantly different from this one but i do see other addons that implement removeHandler.
If that is the case how do i test it? I have added a removeHandler and added some logging but i cant get the logging to happen.

Reply:
I'm not entirely sure myself, but my guess would be to disable a thing, that should remove it's handler instance.

Reply:
@cpmeister So replacing the jar file occasionally calls it. So i am fairly sure this is doing the right thing if it gets called. Strangely for a while i overode unregisterHandler as well and that worked reliably which really doesnt make any sense.
But anyway i think this looks ok now.